File: exercises/python_playwright/Progress_Bar/progress_bar_playwright.py
from playwright.sync_api import Page, expect


class ProgressBarPage:
    def __init__(self, page: Page):
        """Page object for the Progress Bar exercise."""
        self.page = page
        self.start_button = page.get_by_role("button", name="Start")
        self.stop_button = page.get_by_role("button", name="Stop")
        self.progress_bar = page.locator("#progressBar")
        self.result = page.locator("#result")

    def open(self) -> None:
        """Open the Progress Bar exercise page."""
        self.page.goto("http://uitestingplayground.com/progressbar")

    def click_start(self) -> None:
        """Start the progress bar."""
        self.start_button.click()

    def click_stop(self, target: int) -> None:
        """Wait for the progress bar to reach the target, then stop it."""
        handle = self.progress_bar.element_handle()
        self.page.wait_for_function(
            "([element, target]) => Number(element.getAttribute('aria-valuenow')) >= target",
            arg=[handle, target],
            timeout=30000,
            polling=5,
        )
        self.stop_button.click()

File: exercises/javascript_playwright/progress-bar/progressBarPlaywright.js
import { expect } from "@playwright/test";

export class ProgressBarPage {
    /**
     *
     * @param {import('@playwright/test').Page} page
     */
    constructor(page) {
        this.page = page;
        this.startButton = page.locator("#startButton");
        this.stopButton = page.locator("#stopButton");
        this.progressBar = page.locator("#progressBar");
        this.result = page.locator("#result");
    }

    async openExercisePage() {
        await this.page.goto("http://uitestingplayground.com/progressbar");
    }

    async clickStartButton() {
        await this.startButton.click();
    }

    async getProgressBarValue() {
        const value = await this.progressBar.getAttribute("aria-valuenow");
        return Number(value);
    }

    async clickStopButtonAt(percentage) {
        await expect
            .poll(async () => await this.getProgressBarValue(), {
                timeout: 30000,
                intervals: [50, 50, 5, 5],
            })
            .toBe(percentage);
        await this.stopButton.click();
    }

    async getResultText() {
        return await this.result.textContent();
    }
}
